### PR TITLE
Fix debug assertion in BeatFactory::makePreferredBeats()

### PR DESCRIPTION
**Pull request by uklotzde**
Fixes https://github.com/mixxxdj/mixxx/pull/4267#issuecomment-914662378

### CODE REVIEW

File: src/track/beatfactory.cpp [hunk 1]
```diff
@@ -79,15 +79,19 @@ mixxx::BeatsPointer BeatFactory::makePreferredBeats(
         const mixxx::Bpm constBPM = BeatUtils::makeConstBpm(
                 constantRegions, sampleRate, &firstBeat);
         firstBeat = BeatUtils::adjustPhase(firstBeat, constBPM, sampleRate, beats);
-        auto pGrid = mixxx::BeatGrid::makeBeatGrid(
-                sampleRate, constBPM, firstBeat.toNearestFrameBoundary(), subVersion);
-        return pGrid;
+        if (firstBeat.isValid()) {
```
Comment (Holzhaus) [inline src/track/beatfactory.cpp:82]: The method is only called by analyzerbeats. Can we move the check there and leave the debug assertion intact?

**Review by daschuer — APPROVED** [review]

File: src/track/beatfactory.cpp (hunk 1 above)
Comment (uklotzde) [reply src/track/beatfactory.cpp:82]: Sure. I only wonder why the assertion was triggered by firstBeat.toNearestFrameBoundary(), because adjustPhase() already requires a valid value? firstBeat is invalidated by makeConstBpm().

Comment (uklotzde) [reply src/track/beatfactory.cpp:82]: Em, no, I don't think we can't move this check out of BeatFactory::makePreferredBeats(). Will test again.

Also edge cases with a single beat are not handled correctly. It doesn't make any sense to infer a tempo for less than 2 beats.

**Comment by uklotzde** [comment]
Ping.

I suggest to keep the input validation and handling inside the functions instead of making assumptions by whom and how they are called.

The only exception is to assume that all mandatory parameters are valid, e.g. the sample rate.

**Review by daschuer — APPROVED** [review]
LGTM.
@Holzhaus: merge?

**Comment by Holzhaus** [comment]
I'm currently traveling so I cannot retest this. But the code looks good to me. No objections.

**Comment by daschuer** [comment]
Ok than I will hit merge. Thank you.
